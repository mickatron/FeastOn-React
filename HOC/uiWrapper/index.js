import uiWrapper from './uiWrapper';

export default uiWrapper;
