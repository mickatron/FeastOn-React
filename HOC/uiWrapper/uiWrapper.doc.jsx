/* eslint-disable */
import React from 'react';
import PropTypes from 'prop-types';
import asType from '../../proptypes/as';
import classNameType from '../../proptypes/className';

/**
 * The `uiWrapper` function is a Higher Order Component(HOC) that composes some default 
 * functionality and `PropTypes` for a components HTML wrapper element. 
 * 
 * ** Features **
 * * Wrapper element can be changed to any valid HTMl element or React Class.
 * * Custom classNames can be passed in as a string, array or Object in `classnames` module format. 
 * * Props that are not handled by the `uiWrapper` or the component it wraps are spread as
 * JSX props on the `uiWrapper` element, allowing you to set attributes on the wrapper.
 * * BEM modifiers can be applied. the PropTypes values can also be customized.
 * * If a prop passed to the Wrapped Component matches a modifier, the modifier will be applied.
*/
function uiWrapper(props) {
  return <React.Fragment />;
}

uiWrapper.propTypes = {
  /** Allows dynamic rendering of the wrapper element. */
  as: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.string,
    PropTypes.symbol,
  ]),
  /** A custom className. */
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.array,
    PropTypes.object,
  ]),
  /** The Tile components child elements. */
  children: PropTypes.node.isRequired,
  /** BEM modifier names to be applied to the elements wrapper.*/
  modifier: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),
  /** Width className */
  width: PropTypes.oneOf([
    'one-quarter',
    'one-third',
    'one-half',
    'two-thirds',
    'three-quarters',
    'one-whole',
  ]),
};

uiWrapper.defaultProps = {
  as: 'uiWrapper',
  className: undefined,
  modifier: undefined,
  width: undefined,
};

export default uiWrapper;
