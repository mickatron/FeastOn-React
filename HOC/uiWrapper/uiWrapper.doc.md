The `uiWrapper` function is a Higher Order Component(HOC) that composes some default functionality 
and `PropTypes` for a component HTML wrapper element. 


#### Editable modifier Props
You can edit allowable `modifier` PropTypes for a component composed with `uiWrapper`.

```js static
import editModifierProps from '/utils/modifierProps';

editModifierProps('Button', []);

editModifierProps('Button', [], false);
```

#### Editable width Props


### Usage

Accepts a single configuration `object`. The components function signature is below.
```js static
uiWrapper({
  name, // The name of the component
  element = 'div', // the default element to render, can be a string or component.
  Component, // A `React` component the HOC to compose.
  blockName = undefined, // A Block__Element name for the component, if none is specified the name `name` property will be used.
  modifiers = [], // a list of modifiers the component accepts. End user can augment and edit these values.
});
```