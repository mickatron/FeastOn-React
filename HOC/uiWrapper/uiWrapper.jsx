import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

import asType from '../../proptypes/as';
import classNameType from '../../proptypes/className';
import { getWidths } from '../../proptypes/editable/widthProps';
import { getModifiers } from '../../proptypes/editable/modifierProps';

// This function takes a component...
export default function uiWrapperHOC({
  name,
  element = 'div',
  Component,
  blockName = undefined,
  modifiers = [],
}) {
  const modifiersAll = getModifiers(name, modifiers);
  const widths = getWidths();
  const wrapperClassName = blockName || name;
  // ...and returns another component...
  return class UiWrapper extends React.PureComponent {
    static propTypes = {
      aria: PropTypes.object,
      /** The type of element to render the component as. */
      as: asType,
      /** A custom className. */
      className: classNameType,
      /** The Tile components child elements. */
      children: PropTypes.node,
      /** BEM modifier names to be applied to the elements wrapper. */
      modifier: PropTypes.oneOfType([
        PropTypes.oneOf(modifiersAll),
        PropTypes.arrayOf(PropTypes.oneOf(modifiersAll)),
      ]),
      theme: PropTypes.string,
      // TODO: should be renamed: widthClass, span, helpers...
      // width attribute
      /** Width className */
      widthClass: PropTypes.oneOf(widths),
    }

    static defaultProps = {
      as: element,
      className: null,
      children: null,
      modifier: null,
      widthClass: null,
    }
    state = {
      aria: this.props.aria,
    }

    ariaHandler = (newState, merge = true) => {
      const aria = merge ? {...this.state.aria, ...newState} : newState;
      this.setState({aria});
    }

    render() {
      const {
        as,
        className,
        modifier,
        widthClass,
        theme,
        ...other
      } = this.props;
      const { aria } = this.state;
      const Element = as;

      const themeModifier = theme ? `${wrapperClassName}--${theme}` : null;
      // TODO: refactor these elsewhere
      // Modifier classNames
      const modifierClassNames = Array.isArray(modifier)
        ? modifier.map(eachModifier => `${wrapperClassName}--${eachModifier}`)
        : { [`${wrapperClassName}--${modifier}`]: modifier };
      // If one of the wrapped components props has the same name as a modifier set it as a modifier
      Object.keys(other).forEach((prop) => {
        if (modifiers.indexOf(prop) !== -1 && other[prop]) {
          modifierClassNames[`${wrapperClassName}--${prop}`] = true;
        }
      });

      const wrapperProps = {};
      if (Component) {
        const otherKeys = Object.keys(other);
        Object.keys(Component.propTypes).forEach((prop) => {
          if (otherKeys.indexOf(prop) !== -1) {
            wrapperProps[prop] = other[prop];
            delete other[prop];
          }
        });
      }
      // ... and renders the wrapped component with the fresh data!
      // Notice that we pass through any additional props
      return (
        <Element
          className={classnames(
            className, 
            wrapperClassName, 
            themeModifier, 
            modifierClassNames, 
            widthClass
          )}
          {...other}
          {...aria}
        >
          { Component
            ? <Component ariaHandler={ariaHandler} {...wrapperProps} />
            : this.props.children
          }
        </Element>
      );
    }
  };
}
