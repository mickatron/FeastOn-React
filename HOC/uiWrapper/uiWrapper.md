The `uiWrapper` function is a Higher Order Component(HOC) that composes some default functionality 
and `PropTypes` for a component HTML wrapper element. 

** Editable modifier and width Props **

