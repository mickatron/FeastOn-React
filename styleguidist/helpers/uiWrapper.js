module.exports = {
  props: [
    {
      type: {
        name: 'union',
        value: [{ name: 'func' }, { name: 'string' }, { name: 'symbol' }],
      },
      required: false,
      description: 'The type of element to render the component as.',
      defaultValue: { value: '\'ul\'', computed: false },
      tags: {},
      name: 'as',
    },
    {
      type: {
        name: 'union',
        value: [{ name: 'string' }, { name: 'array' }, { name: 'object' }],
      },
      required: false,
      description: 'A custom className.',
      defaultValue: { value: 'undefined', computed: true },
      tags: {},
      name: 'className',
    },
    {
      type: { name: 'node' },
      required: false,
      description: 'The components child elements.',
      tags: {},
      name: 'children',
    },
    {
      type: {
        name: 'union',
        value: [{ name: 'string' }, { name: 'arrayOf', value: { name: 'string' } }],
      },
      required: false,
      description: 'BEM modifier names to be applied to the elements wrapper.',
      defaultValue: { value: 'undefined', computed: true },
      tags: {},
      name: 'modifier',
    },
    {
      type: {
        name: 'enum',
        value:
    [{ value: '\'one-quarter\'', computed: false },
      { value: '\'one-third\'', computed: false },
      { value: '\'one-half\'', computed: false },
      { value: '\'two-thirds\'', computed: false },
      { value: '\'three-quarters\'', computed: false },
      { value: '\'one-whole\'', computed: false }],
      },
      required: false,
      description: 'Width className',
      defaultValue: { value: 'undefined', computed: true },
      tags: {},
      name: 'width',
    },
  ],
};
