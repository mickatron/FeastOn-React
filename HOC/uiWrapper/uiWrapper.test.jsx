import { shallow } from 'enzyme';
import React from 'react';
import uiWrapper from './uiWrapper';

const TestComponent = uiWrapper({
  name: 'TestComponent',
  blockName: 'Test__component',
});

describe('uiWrapper', () => {
  let component = null;

  it('renders', () => {
    component = shallow((
      <TestComponent
        modifier="test"
      >
        <p>Test Component Content.</p>
      </TestComponent>
    ));
    expect(component).toMatchSnapshot();
  });

  it('renders', () => {
    component = shallow((
      <TestComponent>
        <p>Test Component Content.</p>
      </TestComponent>
    ));
    expect(component).toMatchSnapshot();
  });
});
