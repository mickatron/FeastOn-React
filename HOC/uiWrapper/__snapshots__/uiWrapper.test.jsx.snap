// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uiWrapper renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UiWrapper
    as="div"
    className={undefined}
    modifier="test"
    widthClass={undefined}
  >
    <p>
      Test Component Content.
    </p>
  </UiWrapper>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <p>
        Test Component Content.
      </p>,
      "className": "Test__component Test__component--test",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Test Component Content.",
      },
      "ref": null,
      "rendered": "Test Component Content.",
      "type": "p",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <p>
          Test Component Content.
        </p>,
        "className": "Test__component Test__component--test",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Test Component Content.",
        },
        "ref": null,
        "rendered": "Test Component Content.",
        "type": "p",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`uiWrapper renders 2`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UiWrapper
    as="div"
    className={undefined}
    modifier={undefined}
    widthClass={undefined}
  >
    <p>
      Test Component Content.
    </p>
  </UiWrapper>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <p>
        Test Component Content.
      </p>,
      "className": "Test__component",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Test Component Content.",
      },
      "ref": null,
      "rendered": "Test Component Content.",
      "type": "p",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <p>
          Test Component Content.
        </p>,
        "className": "Test__component",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Test Component Content.",
        },
        "ref": null,
        "rendered": "Test Component Content.",
        "type": "p",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
